### DISCUSSION TITLE
Preview mode changes image

### DISCUSSION THREAD

**Discussion by bdulfer** [vladmandic/sdnext#1226] [General]
I tested the new preview mode TAESD and saw that there were slight differences in the generated images compared to Aprox NN which I normally use.
So I checked all preview modes and it turns out that all images are slightly different:

Full:
![full](https://github.com/vladmandic/automatic/assets/130743831/f4a8a980-69ce-4daa-98a3-81bc274a97af)

Aprox NN
![aprox-nn](https://github.com/vladmandic/automatic/assets/130743831/228fdf6d-8d54-44f7-b386-9cc8453842c7)

Aprox Cheap
![aprox-cheap](https://github.com/vladmandic/automatic/assets/130743831/21867738-1d25-48fd-b24c-4617ac7473ce)

TAESD
![taesd](https://github.com/vladmandic/automatic/assets/130743831/7a97982e-f55d-4db5-b050-0fd2d1bd4531)

Just an observation. Nothing serious to investigate, I think.

**Comment by vladmandic** [comment]
Wait, are you saying that preview mode has impact on FINAL image? or just interim steps? Which sampler?

**Reply by bdulfer** [reply]
It seems it was the non-deterministic behaviour of the sampler (DPM++ 2M Karras).

See below.

**Comment by iDeNoh** [comment]
Are the different results consistent? If not couldn't this just be standard non-deterministic behavior?

**Reply by bdulfer** [reply]
I thought of this same question this morning, so I did some tests with different samplers.
To exclude any preview I did them using the API.
There are differences with all samplers but DPM2 Karras, but they are so slight you have to open both images in a viewer and switch back and forth:

**DDIM** (face, claws, tail)
<img src="https://github.com/vladmandic/automatic/assets/130743831/267f95d8-1317-40da-ad97-31202efb6ae7" width="300"> <img src="https://github.com/vladmandic/automatic/assets/130743831/659f0a0c-8f45-4d65-ba83-272f98f537f4" width="300">
<br>
**DPM+ _2M Karras** (wing tip)
<img src="https://github.com/vladmandic/automatic/assets/130743831/86f56c84-5e1c-4e06-a9f6-e1a6fc277e42" width="300"> <img src="https://github.com/vladmandic/automatic/assets/130743831/7bbb91ac-afff-4112-a19f-faed4b70f3d3" width="300">
<br>
**DPM++ SDE** (tail, legs)
<img src="https://github.com/vladmandic/automatic/assets/130743831/a4660b84-5ba0-4b2f-a6b7-71be8193941c" width="300"> <img src="https://github.com/vladmandic/automatic/assets/130743831/94dbe114-adfc-4f91-9890-3f54ca405860" width="300">
<br>
**DPM2 Karras** (none)
<img src="https://github.com/vladmandic/automatic/assets/130743831/b93135e6-0968-4e8b-b175-b8be9bd33367" width="300"> <img src="https://github.com/vladmandic/automatic/assets/130743831/113389ad-cbec-4316-b1d1-2879857cc32c" width="300">
<br>
**Euler** (background tree on the left)
<img src="https://github.com/vladmandic/automatic/assets/130743831/0c8c1057-bcd9-4e47-a4df-b1fcad71c05e" width="300"> <img src="https://github.com/vladmandic/automatic/assets/130743831/474d3b4d-f247-403f-8b99-ab535803e0c7" width="300">

So comparing the preview mode using DPM2 Karras there are no differences:

Left to Right: Full, Aprox-NN, Aproc Cheap, TAESD
<img src="https://github.com/vladmandic/automatic/assets/130743831/b14af62e-98c4-4138-8914-fe7ad38cdb56" width="100"> <img src="https://github.com/vladmandic/automatic/assets/130743831/001ae1bd-ceed-411f-aa28-f998542f4c73" width="100"> <img src="https://github.com/vladmandic/automatic/assets/130743831/317238f9-38b9-4940-82e3-457275bc66d7" width="100"> <img src="https://github.com/vladmandic/automatic/assets/130743831/14f58190-09df-4d73-b951-04ab7c178193" width="100">

I had thought all those samplers were deterministic, or so I had read.
Seems to be wrong. The only really deterministic one is DPM2 Karras.

**Reply by vladmandic** [reply]
Did you set cross-attention to run in deterministic mode? Default is non-deterministic.

**Reply by bdulfer** [reply]
Not sure.
I can't find a setting named 'cross-attention'.

The following settings might fit:
* Cross-attention optimization method: **xFormers** (it's as fast as the others and uses only a fraction of vram
* Cross-attention advanced options: **xFormers enable flash Attention**

**Comment by vladmandic** [comment]
yes, that's it. if you use sdp or xfromers, both run in non-deterministic mode by default.
and enabling checkboxes in second row (advanced options) forces them to run in deterministic mode.

**Reply by bdulfer** [reply]
xFormers enable flash Attention was enabled. So they ran in deterministic mode.

I rechecked it using DPM2 Karras and DPM++ 2M Karras, the former was deterministic and the latter was non-deterministic in the first tests. The new tests shows the same result, independent of xFormers enable flash Attention.

The differences are really small, nothing I personally would be concerned of.

Is it of any interest to do extensive testing here? Otherwise I think you can spend your time better. :-)

BTW: Amazing work you are doing! Really appreciated!

**Reply by vladmandic** [reply]
> Is it of any interest to do extensive testing here? Otherwise I think you can spend your time better. :-)

No, not really, I was just replying on the active thread :)
Personally, I think 100% reproducibility is overrated. Especially when people try to reproduce their old prompts and say "in new version of the app things look slightly different" - that means I'd have to lock down no-changes-in-behavior ever which is just not the point of active development and being on the cutting edge.

> BTW: Amazing work you are doing! Really appreciated!

Thanks!

**Reply by bdulfer** [reply]
Ok.

I'll close this.
